From 75c3b7de1c3cf109f6264acd8879888b3c364325 Mon Sep 17 00:00:00 2001
From: Alain Volmat <alain.volmat@foss.st.com>
Date: Tue, 18 Mar 2025 18:19:02 +0100
Subject: [PATCH] media: stm32: csi: correct data-lanes handling starting from
 1

data-lanes properties should start from 1 instead of 0.
Correct the driver code to complain if a data-lane number is 0
and update driver behavior to match with this numbering.

Change-Id: Ica9e555cdce53d5e4b68b4ab3868a99f4c0c6b2c
Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/445120
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Tested-by: Alain VOLMAT <alain.volmat@st.com>
Reviewed-by: Alain VOLMAT <alain.volmat@st.com>
Domain-Review: Philippe CORNU <philippe.cornu@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Philippe CORNU <philippe.cornu@foss.st.com>
---
 drivers/media/platform/st/stm32/stm32-csi.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

--- a/drivers/media/platform/st/stm32/stm32-csi.c
+++ b/drivers/media/platform/st/stm32/stm32-csi.c
@@ -375,12 +375,12 @@ static int stm32_csi_setup_lane_merger(s
 
 	for (i = 0; i < csi2priv->num_lanes; i++) {
 		/* Check that lane ID is < max number of lane */
-		if (csi2priv->lanes[i] >= STM32_CSI_LANES_MAX) {
+		if (!csi2priv->lanes[i] || csi2priv->lanes[i] > STM32_CSI_LANES_MAX) {
 			dev_err(csi2priv->dev, "Invalid lane id (%d)\n",
 				csi2priv->lanes[i]);
 			return -EINVAL;
 		}
-		lmcfgr |= ((csi2priv->lanes[i] + 1) << ((i * 4) +
+		lmcfgr |= (csi2priv->lanes[i] << ((i * 4) +
 			   STM32_CSI_LMCFGR_DL0MAP_SHIFT));
 	}
 
@@ -506,10 +506,10 @@ static int stm32_csi_start(struct stm32_
 
 	/* Prepare lanes related configuration bits */
 	for (i = 0; i < csi2priv->num_lanes; i++) {
-		if (!csi2priv->lanes[i]) {
+		if (csi2priv->lanes[i] == 1) {
 			lanes_ie |= STM32_CSI_SR1_DL0_ERRORS;
 			lanes_en |= STM32_CSI_PCR_DL0EN;
-		} else {
+		} else if (csi2priv->lanes[i] == 2) {
 			lanes_ie |= STM32_CSI_SR1_DL1_ERRORS;
 			lanes_en |= STM32_CSI_PCR_DL1EN;
 		}
